Usando o Respect (RFC)

augustohp opened the pull request:
Criei este branch apenas pra um estudo e ensaio de como utilizar o Respect mantendo o máximo do código que temos hoje. Gostaria de opiniões para um merge ao **develop**.

Fora os testes eu não fiz mas nenhuma verificação, isto é: as rotas deveriam funcionar com o Respect\Rest.

Existe um refactor que pode definitivamente ser aproveitado (`git cherry-pick`) no  commit 21bd822.

Eu preciso ainda criar uma forma melhor de utilizar regras de validação da nossa _namespace_ com o `Respect\Validation`, pretendo fazer isso nos próximos dias.

Já aproveitei e fiz o merge do branch do **travis** com um **MakeFile** meio crú mas que está funcionando.

[Suite rodando no PHP 5.4 do Travis](https://travis-ci.org/#!/iMastersDev/oportunidades/jobs/3053504)

Commit:
Adiciona Respect/Rest, Respect/Validation e Respect/Config ao composer.json.

Commit:
Refatora `Ophportunidades\DataAccess\PDODataAccess`.

Melhora a cobertura por testes unitários e corrige um bug
encontrado em `PDODataAccess::getAll()` que utilizava uma
classe inexsitente para o método `PdoStatement::setFetchMode`.

PS: Pagando o preço do TDD =P

Commit:
Cria rotas básicas usando o Respect/Rest no `index.php`.

As rotas sao implementadas por duas classes:
    - Ophportunidades/Route/AllPositions.php
    - Ophportunidades/Route/OnePosition.php

São elas (as rotas):
    - [GET] /position           Retorna uma posição
    - [GET] /positions          Retorna todas as posições cadastradas
    - [POST] /position          Cria uma noca posição

Usei o Respect/Config para configurar as dependências das classes
já existentes no `conf/app.ini` e fiz um teste simples te *integration*
só pra garantir que as coisas funcionam como deveriam.

augustohp commented:
Me dêem o feedback nessa bomba que eu já ou removo ou faço o merge dessa birosca aqui :tada: /cc @rogeriopradoj @diegoholiveira

hussani commented:
+1 pro Respect

Por mim pode dar merge

MichelAraujo commented:
+1 pro Respect

augustohp commented:
:sunglasses: @hussani e @MichelAraujo Hoho, Valeu pelo feedback :D

Commit:
Atualiza README.md com info de desenvolvimento.

O README tava um pouco esquecido. Dei uma atualizada bem bunda nele
mas que pelo menos cobre um **workflow de GIT** bem básico pra gente
se organizar e como o resto do pessoal pode fazer pra contribuir com
o desenvolvimento da aplicação.

PS: Eu sei que apesar de lista o Vagrant como requisito do README ele
não está disponível ainda, mas pelo menos isso torna desnecessária a
edição do README quando ele for implementado (espero) num futuro breve.

victorferreira commented:
Já falei que sou a favor do uso do Respect em outra issue mas vai aqui o meu :+1:

diegoholiveira commented:
@augustohp tá ótimo. só faço uma sugestão: o alganet postou no repositório de samples do respect um exemplo de como ele usa o Respect Config para fazer o setup das rotas em um arquivo ini, eu gostei demais desse exemplo dele porque deixou o código muito limpo, bem direto ao ponto. Talvez isso possa ser implementado aqui. O que acha?

victorferreira commented:
@diegoholiveira Dê uma olhada na branch do Respect que o @augustohp criou. Lá está sendo usado o Rest, o Config e o Validation.

augustohp commented:
@diegoholiveira Eu to ligado, tb curti mas esse branch eu tinha feito há muito tempo... To fazendo revisão e melhorando ele ainda. Mas essa semana já rol o merge.
—
Sent from Mailbox for iPad

On Sat, Jul 13, 2013 at 12:15 PM, Victor Ferreira
notifications@github.com wrote:

> ## @diegoholiveira Dê uma olhada na branch do Respect que o @augustohp criou. Lá está sendo usado o Rest, o Config e o Validation.
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/iMastersDev/oportunidades/pull/57#issuecomment-20921262

Commit:
Cria validador e regra para IDs.

- Criado FACADE para permitir regras dentro da nossa namespace.
- Criada regra para validação de IDs.
- Implementada regra para validação de IDs.

Commit:
Cria target `coverage-html` no Makefile.

- Geração de coverage dentro do diretório "reports"
- Diretório "reports" acrescentado ao .gitignore

Commit:
Refactor de bootstrap e rotas com Respect/Config.

- Remove bootstrap alternativo sem autoload do Composer.
- Cria configurações da aplicação baseada em ambientes, tornando o
  antigo conf/app.ini obsoleto.
- Move configurações de rotas do index.php para conf/httpRoutes.ini
  usando o Respect/Config.
- Move configuração de acesso a dados do obsoleto conf/app.ini para
  conf/dataAccess.ini.
- Define configuração de ambiente no Boxfile para o Pagodabox.

O processo de bootstrap agora se divide em 3 etapas:

- Carregamento do composer e declaração de constantes de diretório.
- Declaração de constantes de ambiente e construção do container de
  configuração.
- Configuração da aplicação.

augustohp commented:
Acho melhor fazer o _merge_ da _issue_ #14 antes de mergear esse cara aqui. Óbvio que o Travis me fodeu, mas na minha máquina local ta rolando...

Vou mexer (provavelmente) em algunas coisas ainda, mas um code-review até aqui seria muito bem vindo. :P

rogeriopradoj commented:
Cara, por favor não fica bravo, mas ficou engraçada a frase: [...] mas na minha máquina local ta rolando :-)

![Aqui tá normal](http://vidadesuporte.com.br/wp-content/uploads/2010/05/suporte_50.jpg)

augustohp commented:
@rogeriopradoj Nem me fala! AHUuHAhuAhuahua

danizord commented on composer.json:
```diff
@@ -5,7 +5,11 @@
         "phpunit/phpunit": "3.7@stable",
-        "squizlabs/php_codesniffer": "dev-master"
+        "squizlabs/php_codesniffer": "dev-master",
+        "phpunit/phpunit": "3.7@stable",
```
2? :smile:

augustohp replied:
Pra ter certeza que ele instala né! #not

PS: Fiz o merge desse conflito igual minha bunda. Valeu pelo review!

danizord commented on composer.json:
```diff
@@ -5,7 +5,11 @@
+        "phpunit/phpunit": "3.7@stable",
+        "Respect/Rest": "0.5.1",
+        "Respect/Config": "0.3.1",
+        "Respect/Validation": "0.4.4"
```
Os `Respect/*` deveriam ir pro `require`, né?

danizord commented on public/bootstrap.php:
```diff
@@ -1,33 +1,36 @@
-
 $composer_autoload = APP_ROOT . DS . 'vendor' . DS . 'autoload.php';
-if (!@include($composer_autoload)) {
+if (false === file_exists($composer_autoload)) {
```
`if (!file_exists($composer_autoload)) {` fica mais legível.

augustohp replied:
Blz, é costume. Eu curto os "Yoda Conditionals" :P

danizord commented:
@augustohp tá escavando ae? 😂

augustohp commented:
> @augustohp tá escavando ae? 😂

HAHAHAHA. Descobri o link "Pull Requests" no cabeçalho do GitHub agora. Tem uma penca de coisa velha aberta que eu nem lembrava mais que existia. 😜
